pallets/did/src/lib.rs

```diff
@@ -134,26 +174,290 @@ pub mod pallet {
+
+			// Check DID does not exist yet
+			ensure!(!Did::<T>::contains_key(did.clone()), Error::<T>::DidAlreadyExists);
+
+			// Add assertion method
+			let maybe_assertion_method = if let Some(assertion) = assertion {
+				Some( AssertionMethod::<T> { controller: assertion } )
+			} else { None };
+
+			// Add services
+			let maybe_services = if let Some(services) = services {
+				Some ( Self::try_insert_services(services)? )
+			} else { None };
+
+			// Build Document
+			let document = Document {
 				controller,
-				authenticator,
-				signature
+				authentication: AuthenticationMethod { controller: authentication },
+				assertion_method: maybe_assertion_method,
+				services: maybe_services,
+			};
+
+			// Reserve did deposit. Should not fail as free_balance has been checked
+			T::Currency::reserve(&origin, T::DidDeposit::get());
+
+			// Store new DID
+			Did::<T>::insert(did.clone(), document.clone());
+
+			// Event
+			Self::deposit_event(Event::DidCreated{
+				did,
+				document
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(1000000)]
+		pub fn update_did(
+			origin: OriginFor<T>,
+			did: DidIdentifierOf<T>,
+			controller: Option<DidIdentifierOf<T>>,
+			authentication: Option<T::AuthenticationAddress>,
+			assertion: Option<T::AssertionAddress>,
+			services: Option<BoundedVec<Service<T>, T::MaxServices>>,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+
+			// Try to mutate the document associated with the given DID identifier.
+			Did::<T>::try_mutate(did.clone(), |maybe_doc| -> Result<(), DispatchError> {
+				// If document exists get mutable reference. Otherwise return error
+				let did_doc = maybe_doc.as_mut().ok_or(Error::<T>::DidNotFound)?;
+				// ensure that the caller is the controller of the DID
+				Self::ensure_controller(caller, &did_doc);
+
+				// Try to update the document
+				Self::try_update_did(
+					did_doc,
+					did.clone(),
+					controller,
+					authentication,
+					assertion,
+					services
+				)?;
+
+				// Emit event
+				Self::deposit_event(Event::DidUpdated{
+					did,
+					document: did_doc.clone()
+				});
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000000)]
+		pub fn force_update_did(
+			origin: OriginFor<T>,
+			did: DidIdentifierOf<T>,
+			controller:Option<DidIdentifierOf<T>>,
+			authentication: Option<T::AuthenticationAddress>,
+			assertion: Option<T::AssertionAddress>,
+			services: Option<BoundedVec<Service<T>, T::MaxServices>>,
+		) -> DispatchResult {
+			// Origin ONLY GovernanceOrigin
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			Did::<T>::try_mutate(did.clone(), |maybe_doc| -> Result<(), DispatchError> {
+				let did_doc = maybe_doc.as_mut().ok_or(Error::<T>::DidNotFound)?;
+
+				Self::try_update_did(
+					did_doc,
+					did.clone(),
+					controller,
+					authentication,
+					assertion,
+					services
+				)?;
+				Self::deposit_event(Event::DidForcedUpdated{
+					did,
+					document: did_doc.clone()
+				});
+				Ok(())
```

Comment:
```suggestion
			})?;
			Ok(())
```
Preparing for returning a `DispatchResultWithPostInfo` with another Ok(weight) based on the number of `Services`